### PR TITLE
fix: cleanup defer handling within the QP

### PR DESCRIPTION
**Pull request by dariuszkuc**
If user explicitly opt-out of the `@defer` support, deferred queries will be rejected by the router with GraphQL validation error (unknown directive). It is not possible to plan a query with a `@defer` in it when defer support is disabled so we can simplify this logic.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-federation/tests/query_plan/mod.rs [hunk 1]
```diff
@@ -1,5 +1,3 @@
 #[macro_use]
 mod build_query_plan_support;
 mod build_query_plan_tests;
-// TODO: port query-planner-js/src/__tests__/buildPlan.*.test.ts as new modules here
-mod operation_validations_tests;
```
Comment (dariuszkuc) [inline apollo-federation/tests/query_plan/mod.rs:5]: NOTE: this change is unrelated to the defer cleanup. Those tests were not implemented and their JS counterpart were covering general GraphQL validations that are already handled by `apollo-rs` and reuse fragment logic.

File: apollo-federation/src/operation/tests/mod.rs [hunk 2]
```diff
@@ -17,8 +17,6 @@ use crate::schema::position::InterfaceTypeDefinitionPosition;
 use crate::schema::position::ObjectTypeDefinitionPosition;
 use crate::schema::ValidFederationSchema;
 
-mod defer;
```
Comment (dariuszkuc) [inline apollo-federation/src/operation/tests/mod.rs:20]: those tests were verifying `without_defer` functionality

**Comment by dariuszkuc** [comment]
Was going to remove `QueryPlannerConfig#incremental_delivery` but that is actually used to generate proper API schema. Comment on `QueryPlanIncrementalDeliveryConfig` seems reasonable for "future" use case (i.e. when subgraphs add support for `@defer/@stream`) .... but it might be a while before it happens so I'd just drop it and reintroduce it back WHEN we need it.

We generate API schema twice -> once by the router and once by the QP. Maybe we should update our `QueryPlanner::new` signature to accept `Schema` from the router instead (it contains both the supergraph and api schema)?

File: apollo-federation/src/operation/directive_list.rs [hunk 3]
```diff
@@ -332,6 +335,18 @@ impl DirectiveList {
         inner.rehash();
         Some(item)
     }
+
+    /// Removes @defer directive from self if it has a matching label.
+    pub(crate) fn remove_defer(&mut self, defer_labels: &IndexSet<String>) {
```
Comment (dariuszkuc) [inline apollo-federation/src/operation/directive_list.rs:340]: This was lifted out of the old `DeferFilter`. We don't need to pass schema anymore as at this stage all `@defer` applications will already have labels normalized.

**Comment by dariuszkuc** [comment]
We'll tackle the duplicate API schema generation issue and cleanup the overall API in other PR.

**Review by duckki — APPROVED** [review]
Looks good!

File: apollo-federation/tests/query_plan/build_query_plan_tests/defer.rs [hunk 4]
```diff
@@ -6,71 +6,6 @@ fn config_with_defer() -> QueryPlannerConfig {
     config
 }
 
-#[test]
-fn defer_test_handles_simple_defer_without_defer_enabled() {
```
Comment (TylerBloom) [inline apollo-federation/tests/query_plan/build_query_plan_tests/defer.rs:10]: Is there a reason this was removed? It was ported directly from JS, and, IMO, is testing behavior that we care about.

**Review by TylerBloom — APPROVED** [review]
Other than the one comment about a test being removed, looks good to me.

File: apollo-federation/tests/query_plan/build_query_plan_tests/defer.rs (hunk 4 above)
Comment (duckki) [reply apollo-federation/tests/query_plan/build_query_plan_tests/defer.rs:10]: The changed behavior does not allow `@defer` directive if defer is enabled, while previously `@defer` was removed silently. So, the tested scenario can no longer happen.

However, I just noticed the test would still be able to run. I think this function need to be updated:
https://github.com/apollographql/router/blob/6342406c70edcefbc62bbb6c37bd397a4dc080b9/apollo-federation/tests/query_plan/build_query_plan_support.rs#L98
```rust
pub(crate) fn api_schema_and_planner(
    function_path: &'static str,
    config: QueryPlannerConfig,
    subgraph_names_and_schemas: &[(&str, &str)],
) -> (ValidFederationSchema, QueryPlanner) {
    let supergraph = compose(function_path, subgraph_names_and_schemas);
    let supergraph = apollo_federation::Supergraph::new(&supergraph).unwrap();
    let planner = QueryPlanner::new(&supergraph, config).unwrap();
    let api_schema_config = apollo_federation::ApiSchemaOptions {
        include_defer: true,
        include_stream: false,
    };
    let api_schema = supergraph.to_api_schema(api_schema_config).unwrap();
    (api_schema, planner)
}
```
The config should control `ApiSchemaOptions`'s `include_defer`.

Comment (dariuszkuc) [reply apollo-federation/tests/query_plan/build_query_plan_tests/defer.rs:10]: thats the thing -> outside of tests it is not possible for this scenario to ever reach QP. Old test was verifying the behavior that we would strip out `@defer` from incoming requests if it was disabled.... but if user explicitly opts out of `@defer` but still send a query with `@defer` it will be rejected during GraphQL validation by `apollo-rs` (as that directive won't exist in the API schema).

So real use cases are
- defer enabled
  - query with defer (need to normalize it)
  - query without defer 
- defer disabled
  - query without defer

Comment (dariuszkuc) [reply apollo-federation/tests/query_plan/build_query_plan_tests/defer.rs:10]: Fixed the test api schema config in https://github.com/apollographql/router/pull/6470/commits/6d7a1ca018e9ba35ab8b6937bb4fb9112269352d
